### PR TITLE
fix relay.build to not change the module argument in place

### PR DESCRIPTION
**Pull request by t-vi**
This fixes the observation at
https://discuss.tvm.ai/t/tvm-relay-build-modifying-its-argument/6958

### CODE REVIEW

**Comment by t-vi** [comment]
@tqchen

File: src/relay/backend/build_module.cc [hunk 1]
```diff
@@ -244,6 +244,9 @@ class RelayBuildModule : public runtime::ModuleNode {
       GlobalVar main_glb_var = relay_module->GetGlobalVar("main");
       Function main_func = Downcast<Function>(relay_module->Lookup(main_glb_var));
       auto new_main = BindParamsByName(main_func, params);
+      // copy module to avoid changing our input
+      relay_module = IRModule(relay_module->functions, relay_module->type_definitions,
```
Comment (junrushao1994) [inline src/relay/backend/build_module.cc:248]: If I understand the issue correctly, I think we need to call CopyOnWrite instead.

IRModule is actually defined with `CopyOnWrite` method [here](https://github.com/apache/incubator-tvm/blob/master/include/tvm/ir/module.h#L332), the macro expands to [this](https://github.com/apache/incubator-tvm/blob/master/include/tvm/runtime/object.h#L755-L763). What we need to do is just call its `CopyOnWrite` method to generate a unique copy of IRModule.

Comment (t-vi) [reply src/relay/backend/build_module.cc:248]: OK, so I changed it to call `CopyOnWrite` and it appears to work.

To understand this better, I got the re-instatiation from `FunctionPassNode::operator()` here:
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/ir/transform.cc#L123
is there a reason that isn't done via `CopyOnWrite`?
Superficially, it seems that the update is then 
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/ir/transform.cc#L135
which would seem very similar to the `Update` done in the `Optimize`
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/backend/build_module.cc#L247
which itself just calls add
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/ir/module.cc#L271-L273

Is  something that could also use `CopyOnWrite` or is it not applicable there.

Comment (t-vi) [reply src/relay/backend/build_module.cc]: I must admit I'm not understanding this very well yet, in particular in for the use of the new `relay_module_ptr` I get from `CopyOnWrite` and the original module.
Which one would I use for what after CopyOnWrite?

**Comment by t-vi** [comment]
Now it's not working anymore. :( I think I need a hint.

**Comment by junrushao1994** [comment]
Ooops let me take a closer look

**Comment by tqchen** [comment]
What I would do instead is to simply move `relay_module.CopyOnWrite()` inside `params.size() != 0`, which triggers copy if there is additional copy of relay module outside the function.

Note that for mods pass from python, CopyOnWrite always triggers a copy, unless we do `mod._move()` (moves the python side of the ref)

**Comment by t-vi** [comment]
@tqchen Thank you for the hint. I applied it, but I must admit I'm not entirely certain what's going on.

Do I need to call Update on the ptr returned by CopyOnWrite or on original module? Previously I called it on the original module, but 1) I'm not sure, 2) then the compiler complains about the return of CopyOnWrite being unused.

Also, it seems that CopyOnWrite + Update is certain to do the copy, is it?

**Comment by tqchen** [comment]
We don't have to use the ptr of CopyOnWrite in this case, the module also get updated to points to the new copy .

**Comment by junrushao1994** [comment]
@t-vi Here is what is going on: Internally, `CopyOnWrite` checks if the object (in our case, `relay_module`) is unique. If so, it does nothing; otherwise, it triggers a copy to make sure that the object is unique. Therefore, after calling `CopyOnWrite`, it makes sure that any change only happens locally, in our case, making relay.build don't change the input.

**Comment by t-vi** [comment]
Thanks for the explanation, I'll have to read the source a bit more, but now I can do so with a high-level sense of what's going on. :slightly_smiling_face: 

Happily the CI likes it better now, too.

**Comment by tqchen** [comment]
Thanks @t-vifor the contribution ! Thanks @junrushao1994 for reviewing

**Pull request merged**